lib/Analysis/SecretnessAnalysis/SecretnessAnalysis.cpp

```diff
@@ -109,26 +109,85 @@ LogicalResult SecretnessAnalysis::visitOperation(
+              UnitAttr::get(top->getContext())};
+    }
+    if (secretnessLattice->getValue().getSecretness()) {
+      return {StringAttr::get(top->getContext(),
+                              secret::SecretDialect::kArgSecretAttrName),
+              UnitAttr::get(top->getContext())};
+    }
+    return {StringAttr::get(top->getContext(), "public"),
+            UnitAttr::get(top->getContext())};
+  };
+
   // Add an attribute to the operations to show determined secretness
   top->walk([&](Operation *op) {
-    for (unsigned i = 0; i < op->getNumResults(); ++i) {
-      std::string name = op->getNumResults() == 1
-                             ? "secretness"
-                             : "result_" + std::to_string(i) + "_secretness";
-      auto *secretnessLattice =
-          solver->lookupState<SecretnessLattice>(op->getOpResult(i));
-      if (!secretnessLattice) {
-        op->setAttr(name, StringAttr::get(op->getContext(), "null"));
-        return;
+    // Custom Handling for `func.func`, which uses special attributes
+    if (auto func = llvm::dyn_cast<func::FuncOp>(op)) {
+      // The arguments are either already typed as !secret.secret<..>
+      // or annotated with {secret.secret}, so we don't need to update
+      // them However, the result is generally not, so we need to
+      // annotate that.
+
+      auto *ret = func.getFunctionBody().back().getTerminator();
+      assert(ret->getNumOperands() == func.getFunctionType().getNumResults() &&
+             "Number of returned values does not match function type");
+      for (unsigned i = 0; i < func.getFunctionType().getNumResults(); ++i) {
+        auto *secretnessLattice =
+            solver->lookupState<SecretnessLattice>(ret->getOpOperand(i).get());
+        // We cannot simple add our normal StringAttr or BoolAttr, as
+        // `func.func` only allows dialect attributes as annotations for
+        // arguments and results, so we must use the {secret.secret}
+        // style:
+        if (secretnessLattice &&
+            secretnessLattice->getValue().isInitialized() &&
+            secretnessLattice->getValue().getSecretness()) {
+          func.setResultAttr(
+              i,
+              StringAttr::get(func->getContext(),
+                              secret::SecretDialect::kArgSecretAttrName),
+              UnitAttr::get(func->getContext()));
+        }
       }
-      if (!secretnessLattice->getValue().isInitialized()) {
-        op->setAttr(name, StringAttr::get(op->getContext(), ("unknown")));
-        return;
+    } else if (!llvm::isa<secret::GenericOp>(op)) {
+      // We do not need to add info to secret.generic op returns,
+      // as they already are explicitly typed as !secret.secret<..>
+      SmallVector<NamedAttribute, 1> attributes;
+      if (op->hasTrait<OpTrait::IsTerminator>()) {
+        // Terminators (e.g., func.return, affine.yield) do not have mlir
+        // op results, but do still have logical "results" (mlir operands)
```

Comment:
```suggestion
        // Terminators don't have results, but it is helpful for debugging
        // to show an annotation here. Program analysis will generally not
        // read this, but instead query the op that creates an SSA value
        // for its secretness attribute.
```

IMO Since this is duplicate information that is not primarily intended to be be used by HEIR passes, this comment should reflect that.

Reply:
More generally, I think we should put sufficient documentation to express that this entire annotation routine is meant for debugging. I would prefer we don't have later passes that rely on the annotations, rather than relying on (freshly computed) dataflow analyses. In which case my suggested change to this comment is unnecessary.

Reply:
How about this docstring for the pass
>  Debugging helper that runs the secretness analysis and annotates the IR with the results,
>  extending the `{secret.secret}` annotation to all operation results that are secret. 
>
>  In addition to annotating operation results, the pass also annotates arguments
>  and return types in `func.func` operations, as well as any terminators (e.g. `return`)
> 
>  In `verbose` mode, all results are annotated, including public ones with `{secret.public}`,
>  and values for which the secretness analysis is missing are annotated with `{secret.missing}`,
>  while values where the secretness analysis is inconclusive are annotated with `{secret.unknown}`.

I'd obviously adjust the pass behavior to match

Reply:
Oh, and should the pass also annotate a function arg in the case its type is already `!secret.secret<T>`? Seems a bit redundant.

Reply:
In the case of a pure debugging tool, I don't think it matters. If there were downstream automation relying on the annotations, having a uniform interface would be best. (so no, don't need it)